### PR TITLE
[flang] lower assumed type actual arguments in call statements

### PR DESCRIPTION
**Pull request by cabreraam**
Taking care of TODOs for assumed type actual arguments in `CALL` statements.

The approach here borrows from the commits in [this pull request that addresses assumed type actual arguments in intrinsic functions](https://reviews.llvm.org/D147513).

### CODE REVIEW

File: flang/lib/Lower/ConvertCall.cpp [hunk 1]
```diff
@@ -1172,8 +1173,6 @@ genUserCall(Fortran::lower::PreparedActualArguments &loweredActuals,
       continue;
     }
     const auto *expr = arg.entity->UnwrapExpr();
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertCall.cpp:1175]: If you apply my comment above, you can move this to the MutableBox case where it is used with an assert that this is not null (it is not possible to pass an assumed type to a pointer/allocatable).

File: flang/test/HLFIR/assumed-type-actual-arguments.f90 [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+subroutine test2(x)
+  type(*) :: x
+  interface
+    subroutine fun2(x)
+      type(*) :: x(:)
+    end subroutine
+  end interface
+  call fun2(x)
+end subroutine
```
Comment (jeanPerier) [inline flang/test/HLFIR/assumed-type-actual-arguments.f90:31]: I think this case is illegal. The argument of an rank one deferred shape must be an array (Fortran 15.5.2.4(16)). Not clear to me why this is not enforced in lib/semantics/check-call.cpp like with none assumed type arguments. This may deserve a separate patch with a semantic fix.

File: flang/test/HLFIR/assumed-type-actual-arguments.f90 [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+! CHECK-LABEL: func.func @_QPtest2(
+! CHECK-SAME: %[[ARG0:.*]]: !fir.ref<none> {fir.bindc_name = "x"}) {
+! CHECK:   %[[VAL_0:.*]]:2 = hlfir.declare %[[ARG0]] {uniq_name = "_QFtest2Ex"} : (!fir.ref<none>) -> (!fir.ref<none>, !fir.ref<none>)
+! CHECK:   %[[VAL_1:.*]] = fir.embox %[[VAL_0]]#0 : (!fir.ref<none>) -> !fir.box<none>
+! CHECK:   %[[VAL_2:.*]] = fir.convert %[[VAL_1]] : (!fir.box<none>) -> !fir.box<!fir.array<?xnone>>
+! CHECK:   fir.call @_QPfun2(%[[VAL_2]]) fastmath<contract> : (!fir.box<!fir.array<?xnone>>) -> ()
+! CHECK:   return
```
Comment (jeanPerier) [inline flang/test/HLFIR/assumed-type-actual-arguments.f90:38]: You may want to add some more funny tests.

- passing class(*) to type(*)
- passing optional type(*) deferred shape to optional type(*) assumed size
- passing pointer to optional type(*) deferred shape
- passing rank 2 type(*) deferred shape to rank 1 type(*) assumed size.

**Review by jeanPerier — COMMENTED** [review]
Thank you for working on this!

**Comment by cabreraam** [comment]
Addressed in #83851
